app/frontend/src/features/profile/view/leaveReference/ReferenceForm.tsx

```diff
@@ -0,0 +1,125 @@
+import { Alert } from "@material-ui/lab";
+import { INVALID_STEP } from "features/profile/constants";
+import Appropriate from "features/profile/view/leaveReference/formSteps/Appropriate";
+import Rating from "features/profile/view/leaveReference/formSteps/Rating";
+import SubmitReference from "features/profile/view/leaveReference/formSteps/submit/SubmitReference";
+import Text from "features/profile/view/leaveReference/formSteps/Text";
+import { User } from "pb/api_pb";
+import { ReferenceType } from "pb/references_pb";
+import { useState } from "react";
+import { useRouteMatch } from "react-router-dom";
+import { leaveReferenceBaseRoute, referenceTypeRoute } from "routes";
+import makeStyles from "utils/makeStyles";
+
+export const useReferenceStyles = makeStyles((theme) => ({
+  alert: {
+    marginBottom: theme.spacing(3),
+  },
+  buttonContainer: {
+    display: "flex",
+    justifyContent: "center",
+    paddingTop: theme.spacing(1),
+  },
+  card: {
+    marginTop: theme.spacing(2),
+    marginBottom: theme.spacing(1),
+  },
+  form: {
+    marginBottom: theme.spacing(2),
+  },
+  text: {
+    "& > .MuiInputBase-root": {
+      width: "100%",
+    },
+    marginTop: theme.spacing(1),
+    [theme.breakpoints.up("md")]: {
+      "& > .MuiInputBase-root": {
+        width: 400,
+      },
+    },
+  },
+}));
+
+export type ReferenceContextFormData = {
+  text: string;
+  wasAppropriate: string;
+  rating: number;
+};
+
+export type ReferenceFormInputs = {
+  text: string;
+  wasAppropriate: boolean;
+  rating: number;
+};
+
+export interface ReferenceStepProps {
+  user: User.AsObject;
+  referenceData: ReferenceContextFormData;
+  setReferenceValues: (values: ReferenceContextFormData) => void;
+}
+
+interface ReferenceRouteParams {
+  referenceType: string;
+  userId: string;
+  hostRequest?: string;
+  step?: string;
+}
+
+export interface ReferenceFormProps {
+  user: User.AsObject;
+}
+
+export default function ReferenceForm({ user }: ReferenceFormProps) {
+  const [referenceData, setReferenceData] = useState<ReferenceContextFormData>({
+    text: "",
+    wasAppropriate: "",
+    rating: 0,
+  });
+
+  const setReferenceValues = (values: ReferenceContextFormData) => {
+    setReferenceData((prevData) => ({
+      ...prevData,
+      ...values,
+    }));
+  };
+
+  const friendMatch = useRouteMatch<ReferenceRouteParams>({
+    path: `${leaveReferenceBaseRoute}/:referenceType/:userId/:step?`,
+  });
+  const hostingMatch = useRouteMatch<ReferenceRouteParams>({
+    path: `${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequestId/:step?`,
+  });
+  const allMatch = useRouteMatch<ReferenceRouteParams>({
+    path: `${leaveReferenceBaseRoute}/:referenceType/:userId`,
+  });
```

Comment:
```suggestion
  const leaveReferenceRouteMatch = useRouteMatch<ReferenceRouteParams>({
    path: `${leaveReferenceBaseRoute}/:referenceType/:userId/:hostRequestId/:step?`,
  });
```
Couldn't we just do this in one? `referenceType` and `userId` is going to be available in this still, and then you'd check on the presence of `hostRequestId` to determine if it's a host/surfer reference or a friend's reference. Whenever the hook/any code looks almost identical (especially for three times!), it's always worth thinking about why they look so similar and if you can do less work

Reply:
The reason we didn't do this in one is probably better described in the original conversation about this (lost now in this huge pile of comments), but basically because `hostRequestId` and `step` are both strings it will take the `step` param to be a `hostRequestId`. So for example in the first step, clicking next would take it to i.e. `/leave-reference/friend/:userId/rating` but still display the `Appropriate` component, it now takes `rating` as a `hostRequestId`. Then clicking next again will take it to `/leave-reference/friend/:userId/rating/rating` thereafter.

Reply:
Ahh I see, because of the overloaded routes, my bad for jumping to conclusions too quickly. You could probably still do without the `allMatch` though as that's the same as `friendMatch` but without `step`?